File: services/golang/src/tests/jsonSerde.go
package tests

import (
	"crypto/md5"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

type User struct {
	ID     int    `json:"id"`
	Name   string `json:"name"`
	Email  string `json:"email"`
	Active bool   `json:"active"`
}

type Settings struct {
	Theme         string `json:"theme"`
	Notifications bool   `json:"notifications"`
	Language      string `json:"language"`
}

type Metadata struct {
	Version   string   `json:"version"`
	Timestamp string   `json:"timestamp"`
	Settings  Settings `json:"settings"`
}

type Statistics struct {
	TotalUsers  int      `json:"totalUsers"`
	ActiveUsers int      `json:"activeUsers"`
	AverageAge  float64  `json:"averageAge"`
	Tags        []string `json:"tags"`
}

type SampleData struct {
	Users      []User      `json:"users"`
	Metadata   Metadata    `json:"metadata"`
	Statistics Statistics  `json:"statistics"`
}

func printHash(data interface{}, label string) string {
	jsonBytes, _ := json.Marshal(data)
	hash := md5.Sum(jsonBytes)
	hashHex := fmt.Sprintf("%x", hash)
	if label != "" {
		return fmt.Sprintf("%s MD5 Hash: %s", label, hashHex)
	}
	return fmt.Sprintf("MD5 Hash: %s", hashHex)
}

func JsonSerde(w http.ResponseWriter, r *http.Request) {
	sampleData := SampleData{
		Users: []User{
			{ID: 1, Name: "Alice", Email: "alice@example.com", Active: true},
			{ID: 2, Name: "Bob", Email: "bob@example.com", Active: false},
			{ID: 3, Name: "Charlie", Email: "charlie@example.com", Active: true},
		},
		Metadata: Metadata{
			Version:   "1.0.0",
			Timestamp: time.Now().UTC().Format("2006-01-02T15:04:05.000Z"),
			Settings: Settings{
				Theme:         "dark",
				Notifications: true,
				Language:      "en",
			},
		},
		Statistics: Statistics{
			TotalUsers:  3,
			ActiveUsers: 2,
			AverageAge:  28.5,
			Tags:        []string{"javascript", "nodejs", "benchmark", "json", "serialization"},
		},
	}

	n := 100 // Number of serialization/deserialization cycles

	jsonBytes, _ := json.Marshal(sampleData)
	jsonStr := string(jsonBytes)

	var parsed SampleData
	json.Unmarshal(jsonBytes, &parsed)
	originalHash := printHash(parsed, "Original")

	results := make([]SampleData, 0, n)
	for i := 0; i < n; i++ {
		serialized, _ := json.Marshal(parsed)
		var deserialized SampleData
		json.Unmarshal(serialized, &deserialized)
		results = append(results, deserialized)
	}

	finalHash := printHash(results, fmt.Sprintf("x%d Cycles", n))

	fmt.Fprintf(w, "%s\n%s\nPerformed %d serialization/deserialization cycles\nOriginal data size: %d bytes\nFinal array size: %d objects\n",
		originalHash, finalHash, n, len(jsonStr), len(results))
}

File: services/rust/src/tests/json_serde.rs
use axum::response::Response;
use md5;
use serde::{Deserialize, Serialize};
use serde_json;

#[derive(Serialize, Deserialize, Debug, Clone)]
struct User {
    id: i32,
    name: String,
    email: String,
    active: bool,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
struct Settings {
    theme: String,
    notifications: bool,
    language: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
struct Metadata {
    version: String,
    timestamp: String,
    settings: Settings,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
struct Statistics {
    total_users: i32,
    active_users: i32,
    average_age: f64,
    tags: Vec<String>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
struct SampleData {
    users: Vec<User>,
    metadata: Metadata,
    statistics: Statistics,
}

fn print_hash<T: Serialize>(data: &T, label: &str) -> String {
    let json_str = serde_json::to_string(data).unwrap();
    let hash = md5::compute(json_str.as_bytes());
    let hash_hex = format!("{:x}", hash);
    if !label.is_empty() {
        format!("{} MD5 Hash: {}", label, hash_hex)
    } else {
        format!("MD5 Hash: {}", hash_hex)
    }
}

pub async fn handler() -> Response<String> {
    let sample_data = SampleData {
        users: vec![
            User {
                id: 1,
                name: "Alice".to_string(),
                email: "alice@example.com".to_string(),
                active: true,
            },
            User {
                id: 2,
                name: "Bob".to_string(),
                email: "bob@example.com".to_string(),
                active: false,
            },
            User {
                id: 3,
                name: "Charlie".to_string(),
                email: "charlie@example.com".to_string(),
                active: true,
            },
        ],
        metadata: Metadata {
            version: "1.0.0".to_string(),
            timestamp: chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
            settings: Settings {
                theme: "dark".to_string(),
                notifications: true,
                language: "en".to_string(),
            },
        },
        statistics: Statistics {
            total_users: 3,
            active_users: 2,
            average_age: 28.5,
            tags: vec![
                "javascript".to_string(),
                "nodejs".to_string(),
                "benchmark".to_string(),
                "json".to_string(),
                "serialization".to_string(),
            ],
        },
    };

    let n = 100; // Number of serialization/deserialization cycles

    let json_str = serde_json::to_string(&sample_data).unwrap();

    let parsed: SampleData = serde_json::from_str(&json_str).unwrap();
    let original_hash = print_hash(&parsed, "Original");

    let mut results = Vec::new();
    for _i in 0..n {
        let serialized = serde_json::to_string(&parsed).unwrap();
        let deserialized: SampleData = serde_json::from_str(&serialized).unwrap();
        results.push(deserialized);
    }

    let final_hash = print_hash(&results, &format!("x{} Cycles", n));

    let output = format!(
        "{}\n{}\nPerformed {} serialization/deserialization cycles\nOriginal data size: {} bytes\nFinal array size: {} objects\n",
        original_hash, final_hash, n, json_str.len(), results.len()
    );

    Response::new(output)
}

